File: DroneStrikers-Server/src/rooms/systems/droneStats.ts
import { createDefaultStatsMap, ModifierType, Stat, Stats, StatType } from "src/types/stats";

export class DroneStats {
  private _stats: Map<StatType, Stat>;

  constructor() {
    this._stats = createDefaultStatsMap();
  }

  /**
   * Retrieves the value of the specified stat.
   * @param stat The stat to retrieve.
   * @returns The value of the stat.
   */
  public getValue(stat: StatType): number {
    return this._stats.get(stat)?.value ?? 0;
  }

  /**
   * Adds a modifier to the specified stat.
   * @param stat The stat to add the modifier to.
   * @param type The type of modifier to add.
   * @param value The value of the modifier.
   * @param sourceID The unique ID of the source of the modifier.
   */
  public addModifier(stat: StatType, type: ModifierType, value: number, sourceID: string): void {
    const statObj = this._stats.get(stat);
    if (statObj) {
      statObj.addModifier({ stat, type, value, sourceID });
    }
  }

  /**
   * Removes a modifier from the specified stat.
   * @param stat The stat to remove the modifier from.
   * @param sourceID The unique ID of the source of the modifier.
   */
  public removeModifier(stat: StatType, sourceID: string): void {
    const statObj = this._stats.get(stat);
    if (statObj) {
      statObj.removeModifier(sourceID);
    }
  }
}

File: Drone Strikers/Assets/Scripts/Game/Deprecated/Drone/DroneStats.cs
using System;
using DroneStrikers.Game.Deprecated.Stats;
using UnityEngine;

namespace DroneStrikers.Game.Deprecated.Drone
{
    /// <summary>
    ///     This component manages the stat values for a drone.
    ///     Other components can access these stats via public properties.
    ///     Specific functionality for any given stat should be implemented in a separate component.
    /// </summary>
    public class DroneStats : MonoBehaviour, IStatsProvider
    {
        [SerializeField] private StatTemplateSO _statsTemplate;
        private StatCollection _stats;

        private void Awake()
        {
            if (_statsTemplate == null) throw new Exception("DroneStats: No stats template assigned.");
            _stats = new StatCollection(_statsTemplate);
        }

        public float GetStatValue(StatTypeSO stat) => _stats[stat].Value;

        // --- Upgrades ---
        /// <summary>
        ///     Adds a modifier to the specified stat.
        /// </summary>
        /// <param name="stat"> The StatTypeSO of the stat to modify. </param>
        /// <param name="type"> The type of modifier to add. </param>
        /// <param name="amount"> The value of the modifier. </param>
        /// <param name="source"> The source of the modifier (e.g. an upgrade). Can be null. </param>
        public void AddModifier(StatTypeSO stat, StatModType type, float amount, object source = null) => _stats.AddModifier(stat, type, amount, source);

        /// <summary>
        ///     Adds a flat modifier to the specified stat.
        /// </summary>
        /// <param name="stat"> The StatTypeSO of the stat to modify. </param>
        /// <param name="amount"> The value of the modifier. </param>
        /// <param name="source"> The source of the modifier (e.g. an upgrade). Can be null. </param>
        public void AddFlatModifier(StatTypeSO stat, float amount, object source = null) => _stats.AddModifier(stat, StatModType.Flat, amount, source);

        /// <summary>
        ///     Adds an additive multiplier modifier to the specified stat.
        /// </summary>
        /// <param name="stat"> The StatTypeSO of the stat to modify. </param>
        /// <param name="amount"> The value of the modifier. </param>
        /// <param name="source"> The source of the modifier (e.g. an upgrade). Can be null. </param>
        public void AddAdditiveModifier(StatTypeSO stat, float amount, object source = null) => _stats.AddModifier(stat, StatModType.PercentAdditive, amount, source);

        /// <summary>
        ///     Adds a multiplicative multiplier modifier to the specified stat.
        /// </summary>
        /// <param name="stat"> The StatTypeSO of the stat to modify. </param>
        /// <param name="amount"> The value of the modifier. </param>
        /// <param name="source"> The source of the modifier (e.g. an upgrade). Can be null. </param>
        public void AddMultiplicativeModifier(StatTypeSO stat, float amount, object source = null) => _stats.AddModifier(stat, StatModType.PercentMultiplicative, amount, source);

        /// <summary>
        ///     Removes all modifiers from all stats that originate from the specified source.
        /// </summary>
        /// <param name="source"> The source of the modifiers to remove. </param>
        /// <returns> The total number of modifiers removed across all stats. </returns>
        public int RemoveAllModifiersFromSource(object source) => _stats.RemoveAllModifiersFromSource(source);
    }
}